### PR TITLE
Added n_terms and enrichment_pct calculations to biodomains output

### PR DESCRIPTION
**Pull request by jaclynbeck-sage**
Added three new fields to the output of the biodomains transform:

1. `n_gene_biodomain_terms` is the number of GO terms associated with each Ensembl ID, for each biodomain. 
2. `n_biodomain_terms` is the number of GO terms associated with each biodomain
3. `pct_linking_terms` is the percentage of GO terms for this gene that are in the specified biodomain, defined as (`n_gene_biodomain_terms / n_gene_total_terms * 100`) where `n_gene_total_terms` is the total number of GO terms associated with the gene, across all biodomains. 

The output for each gene now looks like this:
```
  {
    "ensembl_gene_id": "ENSG00000000457",
    "gene_biodomains": [
      {
        "biodomain": "Immune Response",
        "go_terms": [
          "inflammatory response"
        ],
        "n_biodomain_terms": 852,
        "n_gene_biodomain_terms": 1,
        "pct_linking_terms": 33.33
      },
      {
        "biodomain": "Proteostasis",
        "go_terms": [
          "Golgi apparatus",
          "Golgi membrane"
        ],
        "n_biodomain_terms": 695,
        "n_gene_biodomain_terms": 2,
        "pct_linking_terms": 66.67
      }
    ]
  },
...
```

### CODE REVIEW

**Review by BWMac — APPROVED** [review]
LGTM. My only comment (more of a NIT, really) is that it might make sense to get into the habit of breaking up some of this new logic into smaller functions. For example the snippets:

```
totals = (
        genes_biodomains.groupby("biodomain")["go_terms"]
        .nunique().reset_index()
        .rename(columns={"go_terms": "n_go_terms"}))
```

and

```
# Count the number of go_terms associated with each gene
genes_biodomains["n_terms"] = genes_biodomains["go_terms"].apply(len)
# Calculate percent enrichment: merge the list of n_go_terms to match with
# each biodomain and then divide n_terms / n_go_terms
genes_biodomains = genes_biodomains.merge(totals, on="biodomain", how="left")
genes_biodomains["enrichment_pct"] = \
genes_biodomains["n_terms"] / genes_biodomains["n_go_terms"] * 100
# Remove n_go_terms from the data frame
genes_biodomains = genes_biodomains.drop(columns="n_go_terms")
```
    
Could each be smaller functions within `transform_genes_biodomains`.  The reason is to make writing tests easier (smaller function = smaller test). This refactoring work could also be included in the test writing work later.

**Review by JessterB — APPROVED** [review]
The code looks good to me, but given the comment thread on https://sagebionetworks.jira.com/browse/AG-991 we might want to adapt this to use syn51077636 as the source file. 

That said, we could go ahead and merge this for now so that we can load a version of this data with the new fields into dev. We don't have to rush though, as Lawrence won't be starting the biodomains implementation for at least a couple of weeks and possibly longer.

**Comment by jaclynbeck-sage** [comment]
Holding off on merging until I've addressed the re-factor in the AG-991 JIRA task. Then I'll re-submit for review.

**Comment by JessterB** [comment]
1. The new version of the file didn't get uploaded to the Agora Test Data folder from the CI run, it looks like the upload errored out.

From https://github.com/Sage-Bionetworks/agora-data-tools/actions/runs/4189099846/jobs/7261033237

`Either the file path (./staging/genes_biodomains.json) or the destination (syn17015333) are invalid.
400 Client Error: 
Missing 1 part(s).  All parts must be successfully added before a file upload can be completed.`

2. I noticed a `SettingWithCopyWarning` further down at that same link, not sure if that's new or something we should be concerned about.

3.  Are you planning to swap to use syn51077636 instead of syn44151254 as the source file in a separate PR?

**Comment by jaclynbeck-sage** [comment]
@JessterB I'm swapping syn51077636 in for this PR. Working on that now. I'll take a look at why it isn't uploading to the test folder!

**Comment by jaclynbeck-sage** [comment]
I've run this code in a docker container on AWS and it uploads the biodomains file alright. I'm not sure why it failed to do that in the check :( I'll run the check again and see what happens.

**Comment by JessterB** [comment]
It's weird that it only complains about the one file...

**Comment by jaclynbeck-sage** [comment]
It worked the second time. I'm not sure what happened the first time, but the second check I ran uploaded fine.

File: agoradatatools/etl/transform.py
Comment (JessterB) [inline agoradatatools/etl/transform.py:154]: We should probably add a unit test to test_transform.py for this new helper method.

**Comment by jaclynbeck-sage** [comment]
@JessterB @BWMac Re-requesting review after significantly changing the metrics calculations!

**Review by thomasyu888 — APPROVED** [review]
Clean ✨ - thanks for splitting off the `count_grouped_total` function - I'll let Brad comment.

File: agoradatatools/etl/transform.py
Comment (BWMac) [inline agoradatatools/etl/transform.py:158]: Nice!

File: agoradatatools/etl/transform.py
Comment (BWMac) [inline agoradatatools/etl/transform.py:154]: Nice! The other added benefit of this is the extra documentation which will also make developing in and testing this repo easier

**Review by BWMac — APPROVED** [review]

**Review by JessterB — APPROVED** [review]

**Comment by jaclynbeck-sage** [comment]
I've taken a crack at writing a test for the count_grouped_total function. Feedback appreciated!
(All other previously-reviewed code is unchanged, no need to spend time on anything but the new test)

**Review by BWMac — APPROVED** [review]
LGTM

**Review by JessterB — APPROVED** [review]
Thanks Jaclyn!

**Pull request merged**
